fortnite_api/cosmetics/common.py

```diff
@@ -0,0 +1,208 @@
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+from __future__ import annotations
+
+from datetime import datetime
+from typing import Any, Dict, Generic, List, Optional, Tuple, TypeVar
+
+from ..abc import Hashable
+from ..asset import Asset
+from ..enums import CosmeticRarity, CosmeticType
+from ..http import HTTPClientT
+from ..images import Images
+from ..utils import get_with_fallback, parse_time
+
+CosmeticT = TypeVar('CosmeticT', bound='Cosmetic[Any]')
+
+__all__: Tuple[str, ...] = (
+    'Cosmetic',
+    'CosmeticTypeInfo',
+    'CosmeticRarityInfo',
+    'CosmeticSeriesInfo',
+    'CosmeticImages',
+    'CosmeticT',
+)
+
+
+class Cosmetic(Hashable, Generic[HTTPClientT]):
+    """
+    .. attributetable:: fortnite_api.Cosmetic
+
+    Represents a base cosmetic. This class inherits from :class:`~fortnite_api.Hashable`. Every cosmetic type inherits from this class and adds additional attributes.
+    View documentation for the specific cosmetic type for more information.
+
+    - :class:`fortnite_api.CosmeticBr`
+    - :class:`fortnite_api.CosmeticCar`
+    - :class:`fortnite_api.CosmeticInstrument`
+    - :class:`fortnite_api.CosmeticLego`
+    - :class:`fortnite_api.CosmeticLegoKit`
+    - :class:`fortnite_api.CosmeticTrack`
+
+    Attributes
+    ----------
+    id: :class:`str`
+        The ID of the cosmetic.
+    added: :class:`datetime.datetime`
+        When the cosmetic was added.
+    raw_data: :class:`dict`
+        The raw data of the cosmetic.
+    """
+
+    __slots__: Tuple[str, ...] = ('id', 'added', 'raw_data')
+
+    def __init__(
+        self,
+        *,
+        data: Dict[str, Any],
+        http: HTTPClientT,
+    ) -> None:
+        self._http: HTTPClientT = http
+
+        self.id: str = data['id']
+        self.added: datetime = parse_time(data['added'])
+        self.raw_data: Dict[str, Any] = data
+
+
+class CosmeticTypeInfo:
+    """
+    .. attributetable:: fortnite_api.CosmeticTypeInfo
+
+    A class that holds cosmetic type information passed from the API for
+    a given :class:`~fortnite_api.Cosmetic`.
+
+    Attributes
+    ----------
+    value: :class:`fortnite_api.CosmeticType`
+        The value of the cosmetic type.
+    display_value: :class:`str`
+        The display value of the cosmetic type. This is the value that is displayed to the user.
+    backend_value: :class:`str`
+        The internal marker of the cosmetic type.
+    """
+
+    __slots__: Tuple[str, ...] = ('value', 'display_value', 'backend_value')
+
+    def __init__(self, *, data: Dict[str, Any]) -> None:
+        self.value: CosmeticType = CosmeticType(data['value'])
+        self.display_value: str = data['displayValue']
+        self.backend_value: str = data['backendValue']
```

Comment:
I noticed this while implementing the rewrite into my Discord bot, it is cumbersome. I'm torn on the solution for this though.

Conventionally this attr should remain as an enum because it's what the end user is expected to use. This, however, does not get around the issue of obtaining the underlying raw value, which is what many users are going to use. Perhaps a `raw_value` (`str`) attribute should be added for this reason? What do you think

Reply:
Although we discussed potentially reworking the `CosmeticType` enum, this ultimately is not going to work. Not only can an enum not be extended, it could lead to confusion because the enum is also used as some parameters on other API calls. I've exposed the value through a `raw_value` attribute, such that:
```python
val = cosmetic.type.raw_value
```